Major Codebase Refactor for Easier Maintenance

mide opened the pull request:
## Breaking Changes

- Removed Python 2.6 Support. (Also see #41 for this specific issue)
	- `pip` doesn't support 2.6 since fall 2016 https://github.com/pypa/pip/issues/3955
	- `pytest`doesn't support 2.6 since fall 2017  https://github.com/pytest-dev/pytest/issues/2812
	- `setuptools` doesn't support 2.6 since fall 2017 https://github.com/pypa/setuptools/issues/878
	- Many other tools look like they're moving in this direction (and eventually moving to remove all `2.*` support, but I'm not going that far yet.)
- ~~Remove persistent (to file) profiles.~~ 
  - ~~Previously, `aws-google-auth` would write to a user's `~/.aws/config` file and read the values when run the next time. I feel this adds a complexity in trying to configure the tool to determine defaults. `aws-google-auth` supports command line params, user input, environment variables and adding another may be too much to maintain. Of course, we can add this back if desired.~~  (Edit: Per feedback in #38, I will add this back in)


## Under-the-hood Changes

- Broke apart `__init__.py` into the following files (along with their purproses):
  - `google.py` is for all Google related functions. It includes the logic to perform the page scraping and SAML fetching.
  - `amazon.py` for AWS related fucntions. This performs the role extraction from the SAML and performs the AWS API call to get the access tokens.
  - `configuration.py` - This only maintains user options (anything the user specifies). Breaking this into it's own object allows us to perform much more robust testing and just pass around a single object instead of a handful of options.
  - `util.py` for common tooling

## Added the following tests

- [Flake8](https://pypi.python.org/pypi/flake8) Python style testing. This is added into the TravisCI build script (`.travis.yml`).

- `test_configuration.py`:
    - Test that invalid `duration` values get rejected.
    - Test that valid `duration` values are accepted.
    - Test that the default value of `duration` is `max_duration`
    - Test that invalid `ask_role` values get rejected.
    - Test that valid `ask_role` values are accepted.
    - Test that `ask_role` is an optional setting.
    - Test that invalid `idp_id` values get rejected.
    - Test that valid `idp_id` values are accepted.
    - Test that invalid `sp_id` values get rejected.
    - Test that valid `username` values are accepted.
    - Test that invalid `username` values get rejected.
    - Test that valid `sp_id` values are accepted.
    - Test that the default value of `profile` is `sts`
    - Test that invalid `profile` values get rejected.
    - Test that valid `profile` values are accepted.
    - Test all profile_defaults.
    - Test that invalid `region` values get rejected.
    - Test that valid `region` values are accepted.
    - Test that the default value of `region` is `ap_southeast_2`
    - Test that invalid `role_arn` values get rejected.
    - Test that `role_arn_is` is an optional setting.
    - Test that valid `role_arn` values are accepted.
    - Test that invalid `u2f_disabled` values get rejected.
    - Test that valid `u2f_disabled` values are accepted.
    - Test that `u2f_disabled_is` is an optional setting.

- `test_amazon.py`

	- Test that the `sts` `boto` client properly returns an STS client object.
	- Test that role extraction happens correctly given a valid SAML response.
	- Test that role extraction happens correctly given a SAML response with too many commas (See #12)

## Needed Work

In order to get tests to pass, I had to ignore Flake8 [rule E722](https://pycodestyle.readthedocs.io/en/latest/intro.html#error-codes). We should go back and determine what the correct exceptions are to catch and only catch those.
  
## Note
Please don't feel you need to accept this, but do please let me know. If this breaks Cevo's workflows, I may just end up maintaining my own fork.

mide commented:
Wow, only 4.8% increase? That's disappointing.

stevemac007 commented:
I can't see any reason to not accept this work - looks really good, and setup much better to improve the quality of the testing.  Only 4.8% is a good start, and will hopefully allow future enhancements to only improve on that value.

I personally prefer storing the details in the `config` file, makes re-auth easier if you've done it from the input prompts, especially seeing as locating the IDP and SP value in the first place is tricky.

What effort is there for adding that back in?

SamBarker commented:
+1 for maintaining AWS/config support. We have started using google auth with a read only profile and elevating to a rw role as needed, ala sudo. We use AWS-google-auth for the read only login, naming the profile, we then have additional profiles in the config file which are based off the ro one ( we use this between organisations as well)

mide commented:
Sounds good @stevemac007 and @SamBarker - I will add the `~/.aws/config` persistence back in. It shouldn't be too much work.

I do want to clarify though (for @SamBarker) the feature that I did remove (and will put back) is not writing the STS (`aws_access_key_id` and `aws_secret_access_key`) tokens to `~/.aws/credentials`. It's writing various user settings (like email and role_arn) to `~/.aws/config`. 

👍 Thanks for the quick feedback folks. I'll get right on that change.

mide commented:
Implemented feedback; coverage is looking sad, which is a reminder to write tests for the new persistence logic.

mide commented:
Okay, this should be ready for another set of feedback. Please let me know your thoughts.

nonspecialist commented:
Dropping 2.6 is fine so long as we can make sure that people _running_ 2.6 get a Helpful Message which lets them know why the tool stopped working, and what to do about it.

Otherwise, this PR looks _shmick_ (which is a Good Thing)

mide commented:
Okay, good feedback @nonspecialist - I will have it display a note if the user is on Python 2.6.

nonspecialist commented:
Once we merge this, I'd like to get #35 refactored so we can merge it as well

stevemac007 commented:
Test of this branch doesn't seem to read the existing `~/.aws/config` settings.

I have this in my config:
```
[profile cevo-demo]
region = ap-southeast-2
output = json
google_config.role_arn = arn:aws:iam::1234:role/Cevo-Demo
google_config.provider = arn:aws:iam::1234:saml-provider/GoogleApps
google_config.google_idp_id = <IDP>
google_config.google_sp_id = <SPID>
google_config.google_username = steve.mactaggart@email
google_config.duration = 3600
```

But when I run I'm still prompted for my details.

```
$ aws-google-auth -p cevo-demo
Failed to import U2F libraries, U2F login unavailable. Other methods can still continue.
Google username:
```

I'm reading now why this has changed, but seems to have broken backwards compatibility.

mide commented:
@stevemac007 That's likely because I changed the namespace from `google_config` to `aws_google_auth`. But you're right, that did break backwards compatibility. 

I can revert it back to `google_config` all together OR have the application read from `aws_google_auth.*` if its there, and if not, try `google_config`. thoughts?

nonspecialist commented:
hmmm, running in a python2.7 virtualenv and specifying a profile gets me:
```
(aws-google-auth) cmp@tak.local aws-google-auth $ bin/aws-google-auth -p cevo-dev
Failed to import U2F libraries, U2F login unavailable. Other methods can still continue.
Traceback (most recent call last):
  File "bin/aws-google-auth", line 11, in <module>
    load_entry_point('aws-google-auth==0.0.16', 'console_scripts', 'aws-google-auth')()
  File "/home/cmp/Build/Cevo/aws-google-auth/lib/python2.7/site-packages/aws_google_auth-0.0.16-py2.7.egg/aws_google_auth/__init__.py", line 41, in main
    cli(sys.argv[1:])
  File "/home/cmp/Build/Cevo/aws-google-auth/lib/python2.7/site-packages/aws_google_auth-0.0.16-py2.7.egg/aws_google_auth/__init__.py", line 53, in cli
    config.read(args.profile)
  File "/home/cmp/Build/Cevo/aws-google-auth/lib/python2.7/site-packages/aws_google_auth-0.0.16-py2.7.egg/aws_google_auth/configuration.py", line 139, in read
    self.idp_id = util.Util.unicode_to_string_if_needed(util.Util.default_if_none(config_parser[profile].get('aws_google_auth_idp_id', None), self.idp_id))
  File "/home/cmp/Build/Cevo/aws-google-auth/lib/python2.7/site-packages/aws_google_auth-0.0.16-py2.7.egg/aws_google_auth/util.py", line 54, in unicode_to_string_if_needed
    return string.encode('utf-8')
AttributeError: 'NoneType' object has no attribute 'encode'
```

where the ~/.aws/config contains:

```
[cevo-dev]
region = ap-southeast-2
output = json
google_config.role_arn = arn:aws:iam::123456789012:role/Cevo-Dev-Administrator
google_config.provider = arn:aws:iam::123456789012:saml-provider/GoogleApps
google_config.google_idp_id = [elided]
google_config.google_sp_id = [elided]
google_config.google_username = colin.panisset@cevo.com.au
google_config.duration = 3600
```

mide commented:
That's interesting; I'll take a look at that.

nonspecialist commented:
It looks to be because python2.7 doesn't have `string` imported by default

stevemac007 commented:
Happy for it to up convert the config on the fly to the new namespace, but also wary of carrying lots of additional code for the purpose.

Other than it aligns with the tool name, is there great benefit of changing the namespace?

mide commented:
@stevemac007 None - I just figured if a user looked at their `~/.aws/config` file it'd be more obvious what put those values there and what uses them. But it's not important, I can cut it back.

mide commented:
My open tasks:

- [ ] Make message for Python 2.6 users, describing why we did away with 2.6 support.
- [ ] Cut back to `google_config` namespace in `~/.aws/config`. 
- [ ] Look into Python 2.7 errors with `.encode()`.

mide commented:
Alright, implemented feedback. I think we're good for another round. 

I opened https://github.com/cevoaustralia/aws-google-auth/issues/41 so that people that see the message regarding Python 2.6 are given a quick description - so they don't need to read this whole PR.

stevemac007 commented on aws_google_auth/configuration.py:
```diff
@@ -0,0 +1,144 @@
+        config_parser.set(self.profile, 'region', self.region)
+        config_parser.set(self.profile, 'google_config.ask_role', self.ask_role)
+        config_parser.set(self.profile, 'google_config.duration', self.duration)
+        config_parser.set(self.profile, 'google_config.idp_id', self.idp_id)
```
should be `google_config.google_idp_id` for backwards compatibility

mide replied:
Fixed in be0f04e94a1ef33ac65892458d78d21148cc5fd1.

stevemac007 commented on aws_google_auth/configuration.py:
```diff
@@ -0,0 +1,144 @@
+        config_parser.set(self.profile, 'google_config.duration', self.duration)
+        config_parser.set(self.profile, 'google_config.idp_id', self.idp_id)
+        config_parser.set(self.profile, 'google_config.role_arn', self.role_arn)
+        config_parser.set(self.profile, 'google_config.sp_id', self.sp_id)
```
should be `google_config.google_sp_id` for backwards compatibility

mide replied:
Fixed in be0f04e94a1ef33ac65892458d78d21148cc5fd1.

stevemac007 commented on aws_google_auth/configuration.py:
```diff
@@ -0,0 +1,144 @@
+        config_parser.set(self.profile, 'google_config.role_arn', self.role_arn)
+        config_parser.set(self.profile, 'google_config.sp_id', self.sp_id)
+        config_parser.set(self.profile, 'google_config.u2f_disabled', self.u2f_disabled)
+        config_parser.set(self.profile, 'google_config.username', self.username)
```
should be `google_config.google_username` for backwards compatibility

mide replied:
Fixed in be0f04e94a1ef33ac65892458d78d21148cc5fd1.

stevemac007 commented on aws_google_auth/configuration.py:
```diff
@@ -0,0 +1,144 @@
+            self.profile = profile
+            self.ask_role = util.Util.default_if_none(config_parser[profile].getboolean('google_config.ask_role', None), self.ask_role)
+            self.duration = util.Util.default_if_none(config_parser[profile].getint('google_config.duration', None), self.duration)
+            self.idp_id = util.Util.unicode_to_string_if_needed(util.Util.default_if_none(config_parser[profile].get('google_config.idp_id', None), self.idp_id))
```
as above

mide replied:
Fixed in 00c97787b55c0401a450a2f3a1207f720c5d093b.

stevemac007 commented on aws_google_auth/configuration.py:
```diff
@@ -0,0 +1,144 @@
+            self.idp_id = util.Util.unicode_to_string_if_needed(util.Util.default_if_none(config_parser[profile].get('google_config.idp_id', None), self.idp_id))
+            self.region = util.Util.unicode_to_string_if_needed(util.Util.default_if_none(config_parser[profile].get('region', None), self.region))
+            self.role_arn = util.Util.unicode_to_string_if_needed(util.Util.default_if_none(config_parser[profile].get('google_config.role_arn', None), self.role_arn))
+            self.sp_id = util.Util.unicode_to_string_if_needed(util.Util.default_if_none(config_parser[profile].get('google_config.sp_id', None), self.sp_id))
```
as above

mide replied:
Fixed in 00c97787b55c0401a450a2f3a1207f720c5d093b.

stevemac007 commented on aws_google_auth/configuration.py:
```diff
@@ -0,0 +1,144 @@
+            self.role_arn = util.Util.unicode_to_string_if_needed(util.Util.default_if_none(config_parser[profile].get('google_config.role_arn', None), self.role_arn))
+            self.sp_id = util.Util.unicode_to_string_if_needed(util.Util.default_if_none(config_parser[profile].get('google_config.sp_id', None), self.sp_id))
+            self.u2f_disabled = util.Util.default_if_none(config_parser[profile].getboolean('google_config.u2f_disabled', None), self.u2f_disabled)
+            self.username = util.Util.unicode_to_string_if_needed(util.Util.default_if_none(config_parser[profile].get('google_config.username', None), self.username))
```
as above

mide replied:
Fixed in 00c97787b55c0401a450a2f3a1207f720c5d093b.

stevemac007 commented on aws_google_auth/configuration.py:
```diff
@@ -0,0 +1,144 @@
+        profile = util.Util.default_if_none(profile, self.profile)
+        config_parser = configparser.RawConfigParser()
+        config_parser.read(self.config_file)
+        if config_parser.has_section(profile):
```
Looking up the direct profile doesn't work, they should have a `profile` prefix.

Something like:

```
        if profile is None:
            profile_name = "default"
        else:
            profile_name = "profile {}".format(profile)
```

see: https://docs.aws.amazon.com/cli/latest/userguide/cli-multiple-profiles.html

stevemac007 commented on aws_google_auth/configuration.py:
```diff
@@ -0,0 +1,144 @@
+        if amazon_object is not None:
+            credentials_parser = configparser.RawConfigParser()
+            credentials_parser.read(self.credentials_file)
+            if not credentials_parser.has_section(self.profile):
```
Will need to add the `profile` prefix to the non `default` profiles too.

mide replied:
Fixed in 00c97787b55c0401a450a2f3a1207f720c5d093b.

mide commented:
Oh yikes. That’s shameful of me.

I’ll fix the backwards compatibility issues and then add tests to make sure
they stay static.

Sorry for the extra churn.

On Sun, Jan 14, 2018 at 6:47 AM Steve Mactaggart <notifications@github.com>
wrote:

> *@stevemac007* requested changes on this pull request.
>
> Still has a couple of issues with the read/write of the config file format.
> ------------------------------
>
> In aws_google_auth/configuration.py
> <https://github.com/cevoaustralia/aws-google-auth/pull/38#discussion_r161394691>
> :
>
> > +    # regular AWS tooling (aws cli and boto) to use the credentials in the
> +    # profile the user specified.
> +    def write(self, amazon_object):
> +        self.ensure_config_files_exist()
> +
> +        assert (self.profile is not None), "Can not store config/credentials if the AWS_PROFILE is None."
> +
> +        # Write to the configuration file
> +        config_parser = configparser.RawConfigParser()
> +        config_parser.read(self.config_file)
> +        if not config_parser.has_section(self.profile):
> +            config_parser.add_section(self.profile)
> +        config_parser.set(self.profile, 'region', self.region)
> +        config_parser.set(self.profile, 'google_config.ask_role', self.ask_role)
> +        config_parser.set(self.profile, 'google_config.duration', self.duration)
> +        config_parser.set(self.profile, 'google_config.idp_id', self.idp_id)
>
> should be google_config.google_idp_id for backwards compatibility
> ------------------------------
>
> In aws_google_auth/configuration.py
> <https://github.com/cevoaustralia/aws-google-auth/pull/38#discussion_r161394697>
> :
>
> > +    def write(self, amazon_object):
> +        self.ensure_config_files_exist()
> +
> +        assert (self.profile is not None), "Can not store config/credentials if the AWS_PROFILE is None."
> +
> +        # Write to the configuration file
> +        config_parser = configparser.RawConfigParser()
> +        config_parser.read(self.config_file)
> +        if not config_parser.has_section(self.profile):
> +            config_parser.add_section(self.profile)
> +        config_parser.set(self.profile, 'region', self.region)
> +        config_parser.set(self.profile, 'google_config.ask_role', self.ask_role)
> +        config_parser.set(self.profile, 'google_config.duration', self.duration)
> +        config_parser.set(self.profile, 'google_config.idp_id', self.idp_id)
> +        config_parser.set(self.profile, 'google_config.role_arn', self.role_arn)
> +        config_parser.set(self.profile, 'google_config.sp_id', self.sp_id)
>
> should be google_config.google_sp_id for backwards compatibility
> ------------------------------
>
> In aws_google_auth/configuration.py
> <https://github.com/cevoaustralia/aws-google-auth/pull/38#discussion_r161394707>
> :
>
> > +
> +        assert (self.profile is not None), "Can not store config/credentials if the AWS_PROFILE is None."
> +
> +        # Write to the configuration file
> +        config_parser = configparser.RawConfigParser()
> +        config_parser.read(self.config_file)
> +        if not config_parser.has_section(self.profile):
> +            config_parser.add_section(self.profile)
> +        config_parser.set(self.profile, 'region', self.region)
> +        config_parser.set(self.profile, 'google_config.ask_role', self.ask_role)
> +        config_parser.set(self.profile, 'google_config.duration', self.duration)
> +        config_parser.set(self.profile, 'google_config.idp_id', self.idp_id)
> +        config_parser.set(self.profile, 'google_config.role_arn', self.role_arn)
> +        config_parser.set(self.profile, 'google_config.sp_id', self.sp_id)
> +        config_parser.set(self.profile, 'google_config.u2f_disabled', self.u2f_disabled)
> +        config_parser.set(self.profile, 'google_config.username', self.username)
>
> should be google_config.google_username for backwards compatibility
> ------------------------------
>
> In aws_google_auth/configuration.py
> <https://github.com/cevoaustralia/aws-google-auth/pull/38#discussion_r161394714>
> :
>
> > +    # Read from the configuration file and override ALL values currently stored
> +    # in the configuration object. As this is potentially destructive, it's
> +    # important to only run this in the beginning of the object initialization.
> +    # We do not read AWS credentials, as this tool's use case is to obtain
> +    # them.
> +    def read(self, profile):
> +        self.ensure_config_files_exist()
> +
> +        profile = util.Util.default_if_none(profile, self.profile)
> +        config_parser = configparser.RawConfigParser()
> +        config_parser.read(self.config_file)
> +        if config_parser.has_section(profile):
> +            self.profile = profile
> +            self.ask_role = util.Util.default_if_none(config_parser[profile].getboolean('google_config.ask_role', None), self.ask_role)
> +            self.duration = util.Util.default_if_none(config_parser[profile].getint('google_config.duration', None), self.duration)
> +            self.idp_id = util.Util.unicode_to_string_if_needed(util.Util.default_if_none(config_parser[profile].get('google_config.idp_id', None), self.idp_id))
>
> as above
> ------------------------------
>
> In aws_google_auth/configuration.py
> <https://github.com/cevoaustralia/aws-google-auth/pull/38#discussion_r161394716>
> :
>
> > +    # We do not read AWS credentials, as this tool's use case is to obtain
> +    # them.
> +    def read(self, profile):
> +        self.ensure_config_files_exist()
> +
> +        profile = util.Util.default_if_none(profile, self.profile)
> +        config_parser = configparser.RawConfigParser()
> +        config_parser.read(self.config_file)
> +        if config_parser.has_section(profile):
> +            self.profile = profile
> +            self.ask_role = util.Util.default_if_none(config_parser[profile].getboolean('google_config.ask_role', None), self.ask_role)
> +            self.duration = util.Util.default_if_none(config_parser[profile].getint('google_config.duration', None), self.duration)
> +            self.idp_id = util.Util.unicode_to_string_if_needed(util.Util.default_if_none(config_parser[profile].get('google_config.idp_id', None), self.idp_id))
> +            self.region = util.Util.unicode_to_string_if_needed(util.Util.default_if_none(config_parser[profile].get('region', None), self.region))
> +            self.role_arn = util.Util.unicode_to_string_if_needed(util.Util.default_if_none(config_parser[profile].get('google_config.role_arn', None), self.role_arn))
> +            self.sp_id = util.Util.unicode_to_string_if_needed(util.Util.default_if_none(config_parser[profile].get('google_config.sp_id', None), self.sp_id))
>
> as above
> ------------------------------
>
> In aws_google_auth/configuration.py
> <https://github.com/cevoaustralia/aws-google-auth/pull/38#discussion_r161394720>
> :
>
> > +    def read(self, profile):
> +        self.ensure_config_files_exist()
> +
> +        profile = util.Util.default_if_none(profile, self.profile)
> +        config_parser = configparser.RawConfigParser()
> +        config_parser.read(self.config_file)
> +        if config_parser.has_section(profile):
> +            self.profile = profile
> +            self.ask_role = util.Util.default_if_none(config_parser[profile].getboolean('google_config.ask_role', None), self.ask_role)
> +            self.duration = util.Util.default_if_none(config_parser[profile].getint('google_config.duration', None), self.duration)
> +            self.idp_id = util.Util.unicode_to_string_if_needed(util.Util.default_if_none(config_parser[profile].get('google_config.idp_id', None), self.idp_id))
> +            self.region = util.Util.unicode_to_string_if_needed(util.Util.default_if_none(config_parser[profile].get('region', None), self.region))
> +            self.role_arn = util.Util.unicode_to_string_if_needed(util.Util.default_if_none(config_parser[profile].get('google_config.role_arn', None), self.role_arn))
> +            self.sp_id = util.Util.unicode_to_string_if_needed(util.Util.default_if_none(config_parser[profile].get('google_config.sp_id', None), self.sp_id))
> +            self.u2f_disabled = util.Util.default_if_none(config_parser[profile].getboolean('google_config.u2f_disabled', None), self.u2f_disabled)
> +            self.username = util.Util.unicode_to_string_if_needed(util.Util.default_if_none(config_parser[profile].get('google_config.username', None), self.username))
>
> as above
> ------------------------------
>
> In aws_google_auth/configuration.py
> <https://github.com/cevoaustralia/aws-google-auth/pull/38#discussion_r161394918>
> :
>
> > +            credentials_parser.set(self.profile, 'aws_session_token', amazon_object.session_token)
> +            with open(self.credentials_file, 'w+') as f:
> +                credentials_parser.write(f)
> +
> +    # Read from the configuration file and override ALL values currently stored
> +    # in the configuration object. As this is potentially destructive, it's
> +    # important to only run this in the beginning of the object initialization.
> +    # We do not read AWS credentials, as this tool's use case is to obtain
> +    # them.
> +    def read(self, profile):
> +        self.ensure_config_files_exist()
> +
> +        profile = util.Util.default_if_none(profile, self.profile)
> +        config_parser = configparser.RawConfigParser()
> +        config_parser.read(self.config_file)
> +        if config_parser.has_section(profile):
>
> Looking up the direct profile doesn't work, they should have a profile
> prefix.
>
> Something like:
>
>         if profile is None:
>             profile_name = "default"
>         else:
>             profile_name = "profile {}".format(profile)
>
> see:
> https://docs.aws.amazon.com/cli/latest/userguide/cli-multiple-profiles.html
> ------------------------------
>
> In aws_google_auth/configuration.py
> <https://github.com/cevoaustralia/aws-google-auth/pull/38#discussion_r161394931>
> :
>
> > +        config_parser.set(self.profile, 'region', self.region)
> +        config_parser.set(self.profile, 'google_config.ask_role', self.ask_role)
> +        config_parser.set(self.profile, 'google_config.duration', self.duration)
> +        config_parser.set(self.profile, 'google_config.idp_id', self.idp_id)
> +        config_parser.set(self.profile, 'google_config.role_arn', self.role_arn)
> +        config_parser.set(self.profile, 'google_config.sp_id', self.sp_id)
> +        config_parser.set(self.profile, 'google_config.u2f_disabled', self.u2f_disabled)
> +        config_parser.set(self.profile, 'google_config.username', self.username)
> +        with open(self.config_file, 'w+') as f:
> +            config_parser.write(f)
> +
> +        # Write to the credentials file (only if we have credentials)
> +        if amazon_object is not None:
> +            credentials_parser = configparser.RawConfigParser()
> +            credentials_parser.read(self.credentials_file)
> +            if not credentials_parser.has_section(self.profile):
>
> Will need to add the profile prefix to the non default profiles too.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cevoaustralia/aws-google-auth/pull/38#pullrequestreview-88680131>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAhW1tt5K3ChgKaFGkp457-l1CWl1nJcks5tKek8gaJpZM4RWd6w>
> .
>
-- 
Sent from my mobile device.

mide commented:
Okay:

- I've addressed the backwards compatibility concerns in be0f04e94a1ef33ac65892458d78d21148cc5fd1, and added tests to ensure things don't change unexpectedly. 
- I fixed the whole named profile issue, and added tests in 00c97787b55c0401a450a2f3a1207f720c5d093b. 

My apologies for the extra churn here, I think we're getting closer.

nonspecialist commented:
Hmmm, I have a profile in `~/.aws/config` with the `profile` prefix in the name; e.g.

```
[profile None]
region = ap-southeast-2
output = json
...
```

It's quite possible that this was due to an intermediate version, but the result is confusing. It breaks with:

```
(aws-google-auth) cmp@tak.local aws-google-auth $ bin/aws-google-auth 
Failed to import U2F libraries, U2F login unavailable. Other methods can still continue.
Google Password: 
Invalid parameters.
Traceback (most recent call last):
  File "bin/aws-google-auth", line 11, in <module>
    load_entry_point('aws-google-auth==0.0.16', 'console_scripts', 'aws-google-auth')()
  File "/home/cmp/Build/Cevo/aws-google-auth/lib/python2.7/site-packages/aws_google_auth-0.0.16-py2.7.egg/aws_google_auth/__init__.py", line 52, in main
    cli(sys.argv[1:])
  File "/home/cmp/Build/Cevo/aws-google-auth/lib/python2.7/site-packages/aws_google_auth-0.0.16-py2.7.egg/aws_google_auth/__init__.py", line 107, in cli
    config.raise_if_invalid()
  File "/home/cmp/Build/Cevo/aws-google-auth/lib/python2.7/site-packages/aws_google_auth-0.0.16-py2.7.egg/aws_google_auth/configuration.py", line 72, in raise_if_invalid
    assert (self.profile.__class__ is str), "Expected profile to be a string. Got {}.".format(self.profile.__class__)
AssertionError: Expected profile to be a string. Got <type 'NoneType'>.
```

If I edit out the `profile` prefix (so it's just `[None]`) then all's well.

If I specify a profile on the command-line, eg `aws-google-auth -p cevo-dev` then it works too; but not if I just run it without setting a default profile.

`$AWS_PROFILE` is unset.

I think it's because `config.read(args.profile)` gets a NoneType first, before any sanity checking is done and then, after environment variables are read again, it's overridden by `default_if_none` which is passed a NoneType in _as the default_

Adding `default='default'` to line 27 of `__init__.py` fixes this but it would be useful to report on _which_ profile is having issues, so the user can adjust/edit their `~/.aws/config` file if needed.

nonspecialist commented:
I think my description is confusing, so I'll adjust it.

Missing any default value for the `profile` argument results in a `NoneType` being used to load, and re-load the profile, instead of using the profile "default"

mide commented:
Hmmm, okay. That's an interesting case. I'll take a look at that in the morning @nonspecialist. _(I'm in the [US - New York time zone](https://duckduckgo.com/?q=current+time+in+Boston))_

mide commented:
@nonspecialist I was able to reproduce the problem and I feel that 6bcc759 should solve your problem. That being said, I don't think we want the default to be `None` and instead be `sts`. So now when you don't specify a profile, it should use `sts`.

SamBarker commented:
@mide sorry for the slow reply. 

Being back in front of the computer again I've just outlined our cross account usage a bit more on https://github.com/cevoaustralia/aws-google-auth/issues/42#issuecomment-358106519 hopefully that makes our current use case a bit clearer.

mide commented:
@SamBarker - I think I'm a little lost, is that just for context, or do you expect changes from that comment?

SamBarker commented:
Just for context! Not expecting anything else at all.

On Wed, 17 Jan 2018 at 10:18, Mark Ide <notifications@github.com> wrote:

> @SamBarker <https://github.com/sambarker> - I think I'm a little lost, is
> that just for context, or do you expect changes from that comment?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cevoaustralia/aws-google-auth/pull/38#issuecomment-358109307>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACs8qkgdhcqIP9nf6u8oXi8t3NJYm1oDks5tLRIxgaJpZM4RWd6w>
> .
>
-- 

[image: MahiFX]

Sam Barker

Senior Developer

------------------------------

MahiFX

Level 3,

50 Victoria Street,

Christchurch, 8013

------------------------------

Mobile: +64 (0)21 1917157

Skype: sammahifx



[image: MFX Compass]

-- 


------------------------------

*IMPORTANT NOTICE*: MahiFX Limited (*MahiFX Ltd*) and MahiFX (UK) Limited (*MahiFX 
(UK)*) are operating subsidiaries within the MahiFX group of companies 
(collectively, the MahiFX Group). All references to “*MahiFX*” refer to the 
MahiFX Group. MahiFX Limited is registered in New Zealand (Company no. 
2446590, NZBusNo 9429031595070) and Australia (Australian registered body 
number ARBN 152-535-085). MahiFX Limited is authorised and regulated under 
the Australian Securities and Investment Commission (AFSL number 414198) 
and the New Zealand Financial Markets Authority (FSPR number FSP197465). 
MahiFX (UK) Limited is registered in the United Kingdom, (registered 
company number 08107062). MahiFX (UK) Limited is authorised and regulated 
under the Financial Conduct Authority (reference number 751019).


This email, its attachments and any rights attaching hereto are 
confidential and intended exclusively for the person to whom the email is 
addressed. If you are not the intended recipient, do not read, copy, 
disclose or use the contents in any way. Please notify the sender by return 
email and destroy the email and attachments immediately. MahiFX does not 
accept any liability for any changes made to this email or attachments 
after sending by MahiFX. You must scan this email and attachments for 
viruses. The opinions expressed are not necessarily those of MahiFX. MahiFX 
accepts no liability for any loss, damage or consequence, whether caused by 
our own negligence or not, resulting directly or indirectly from the use of 
this email and attachments.


For more information about MahiFX Limited see mahifx.com.

nonspecialist commented:
Excellent! Works for my bizarre-o `~/.aws/config` with and without specifying a profile. I like the `coalesce` cleanup too.

👍

mide commented:
Thanks!

stevemac007 commented:
Looks great @mide - I've confirmed that my existing config works well.
Lets get this merged and then we can look at updating the remaining open PR's.

cliveza commented:
@stevemac007 Could we get a new python package please?

cliveza commented:
@stevemac007 Could we get a new python package please?

nonspecialist commented:
@cliveza the travis build should publish a new package automatically, but TravisCI has had some problems over the past 24 hours. I'll check the build and kick it manually if need be

nonspecialist commented:
@cliveza pypi and Docker tags have been updated
